c++/src/TypeImpl.cc

```diff
@@ -491,4 +491,190 @@ namespace orc {
     return std::unique_ptr<Type>(result);
   }
 
+  Type * Type::buildTypeFromString(const std::string& input) {
```

Comment:
This should probably be factored in to smaller functions. Take a look at the equivalent java https://github.com/apache/orc/blob/master/java/core/src/java/org/apache/orc/TypeDescription.java#L396 and you'll see parseName, parseCategory, parseInteger, etc.